lean/bitvec.lean

```diff
@@ -0,0 +1,325 @@
+/-
+Copyright (c) 2015 Joe Hendrix. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joe Hendrix, Sebastian Ullrich, Jason Dagit
+
+Basic operations on bitvectors.
+
+This is a work-in-progress, and contains additions to other theories.
+-/
+import data.vector
+import .nat.lemmas
+
+import tactic.find
+
+import .tactic -- for dec_trivial_tac
+
+-- A `bitvec n` is a subtype of natural numbers such that the value of
+-- the bitvec is strictly less than 2^n.
+@[reducible]
+def bitvec (sz:ℕ) :=
+  { x : ℕ // x < (2 ^ sz) }
+
+namespace bitvec
+open nat
+open int
+open vector
+
+section zero
+
+  -- Create a zero bitvector
+  protected
+  def zero (n : ℕ) : bitvec n :=
+    ⟨0, nat.pos_pow_of_pos n dec_trivial⟩
+
+  -- bitvectors have a zero, at every length
+  instance {n:ℕ} : has_zero (bitvec n) := ⟨bitvec.zero n⟩
+
+  @[simp]
+  lemma bitvec_zero_zero (x : bitvec 0) : x.val = 0 :=
+    begin
+      cases x,
+      { simp [nat.pow_zero, lt_one_is_zero] at x_property,
+        simp, assumption
+      }
+    end
+
+end zero
+
+section one
+
+  -- Create a bitvector with the constant one, doing so requires
+  -- the bitvector to have positive length.
+  protected
+  def one_of_pos_len (n: ℕ) (H : n > 0) : bitvec n :=
+    ⟨1, calc
+          1   < 2^1 : by dec_trivial_tac
+          ... ≤ 2^n : by apply (pow_le_pow_of_le_right (zero_lt_succ _) (succ_le_of_lt H))⟩
```

Comment:
This is personal preference, but I'd probably make the theorem explicit, and move the bitvec assignment around, e.g.,

```
 def one_le_2_n {n: ℕ} (h : n > 0) : 1 < 2^n :=
     calc 1   < 2^1 : by dec_trivial_tac
          ... ≤ 2^n : nat.pow_le_pow_of_le_right (by dec_trivial_tac) h

protected def one : Π(n:ℕ), bitvec n
| 0 := bitvec.zero 0
| (nat.succ n) := ⟨1, one_le_2_n (nat.zero_lt_succ _) ⟩
```

This way the theorem being proven is quite clear.